### PR TITLE
Model bootstrap wireup

### PR DESCRIPTION
**Pull request by tlm**
Wire up model creation process to bootstrap.
    
This PR aims to start wiring up the full model creation process within Dqlite.
    
    - Renamed the cloud and credential bootstrap commands to no longer
      include statements like insert initial controller x. This is because
      their purpose is to seed cloud and credentials. They have no
      specific part that relates to controllers. But they can be used to
      insert both the controller's cloud and other clouds.
    - We now only insert a controller credential if one has been provided.
    - Sets up the controllers model in Dqlite and associates both the type
      of model, cloud, owner and credential.
    - Removed the column from Dqlite that indicates if a cloud is the
      "controller cloud". We can answer this question by pure relational
      data. We also don't have the chance of it being set twice now as we
      can't place a unique index on this column.
    - Have renamed the domain/credential types file from credential.go to
      types.go. This falls in line with what we have done in other domains.
    - Strengthened the concepts around a credential.Id. This type can now
      validate itself for correctly set values and also indicate if it's at
      its zero initialisation value. I have opted to leave this type in the
      domain package as it contains business logic around what a valid
      credential is and isn't.
    - Utility functions provided to transfer from a credential tag to a
      credential.Id.
    - Introduces a new model.AlreadyExists error to better indicate where
      conflicts are happening in the services and state layers.
    - Attempting to document the contracts of services interfaces better so
      we can better write tests for those contracts to assert breakage.
    - New state model creation process that aims to not only register a
      model now but also associate the model's metadata and cloud information
      all within the one transaction.
    - Introduced a new model domain type for describing the arguments
      required to the model creation process.
    - Model domain now contains a function for generating model UUID's. Have
      decided to move this process over to the domain as it's business logic
      around what a model UUID is.
    - Removed an accidental usage of pkg/errors.

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [x] Go unit tests, with comments saying what you're testing
- ~[ ] [Integration tests](https://github.com/juju/juju/tree/main/tests), with comments saying what you're testing~
- [ ] [doc.go](https://discourse.charmhub.io/t/readme-in-packages/451) added or updated in changed packages

## QA steps

Unit tests

## Documentation changes

N/A

## Links

**Jira card:** JUJU-4648

### CODE REVIEW

File: domain/model/bootstrap/bootstrap.go [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+	"fmt"
+
+	"github.com/juju/juju/core/database"
+	"github.com/juju/juju/domain/model"
+	"github.com/juju/juju/domain/model/state"
+)
+
+// Create is responsible for making a new model with all of it's associated
+// metadata during the bootstrap process.
+func CreateModel(
```
Comment (tlm) [inline domain/model/bootstrap/bootstrap.go:18]: Write tests and better document error interface.

File: domain/model/types.go [hunk 2]
```diff
@@ -4,13 +4,111 @@
...
+	// Must not be empty for a valid struct.
+	Cloud string
+
+	// CloudRegion is the region that the model will use in the cloud.
+	// Must not be empty for a valid struct.
+	CloudRegion string
+
+	// Credential is the id attributes for the credential to be associated with
+	// model. Credential must be for the same cloud as that of the model.
+	// Credential can be the zero of the struct to not have a credential
```
Comment (tlm) [inline domain/model/types.go:28]: zero value

File: domain/modelmanager/bootstrap/bootstrap.go [hunk 3]
```diff
@@ -6,18 +6,23 @@ package bootstrap
...
 
 	"github.com/juju/juju/core/database"
 	"github.com/juju/juju/domain/model"
-	modelmanagerstate "github.com/juju/juju/domain/modelmanager/state"
+	"github.com/juju/juju/domain/modelmanager/state"
 )
 
-// InsertModel is a bootstrap convenience function for inserting a new model into
-// the controllers model list.
-func InsertModel(uuid model.UUID) func(context.Context, database.TxnRunner) error {
```
Comment (tlm) [inline domain/modelmanager/bootstrap/bootstrap.go:17]: I think this can be worded a bit better with these changes.

File: domain/schema/controller.go [hunk 4]
```diff
@@ -318,21 +317,47 @@ CREATE TABLE model_list (
...
+
+INSERT INTO model_type VALUES
+    (0, 'none'),
+    (1, 'iaas'),
+    (2, 'caas');
+
 CREATE TABLE model_metadata (
-    model_uuid TEXT PRIMARY KEY,
-    cloud_uuid TEXT,
+    model_uuid            TEXT PRIMARY KEY,
```
Comment (tlm) [inline domain/schema/controller.go:334]: Format this better.

File: juju/testing/apiserver.go [hunk 5]
```diff
@@ -152,7 +153,7 @@ func (noopRegisterer) Unregister(prometheus.Collector) bool {
 	return true
 }
 
-func leaseManager(controllerUUID string, db database.DBGetter, clock clock.Clock) (*lease.Manager, error) {
+func leaseManager(controllerUUID string, db coredatabase.DBGetter, clock clock.Clock) (*lease.Manager, error) {
```
Comment (tlm) [inline juju/testing/apiserver.go:156]: Fix this import up.

File: juju/testing/apiserver.go [hunk 6]
```diff
@@ -383,10 +384,10 @@ func (s *ApiServerSuite) TearDownTest(c *gc.C) {
 
 // SeedControllerCloud is responsible for applying the controller cloud to
 // the given database.
-func (s *ApiServerSuite) SeedControllerCloud(c *gc.C, runner database.TxnRunner) {
+func (s *ApiServerSuite) SeedControllerCloud(c *gc.C, runner coredatabase.TxnRunner) {
```
Comment (tlm) [inline juju/testing/apiserver.go:387]: Fix this import up.

File: juju/testing/apiserver.go [hunk 7]
```diff
@@ -565,19 +566,19 @@ func (s *ApiServerSuite) SeedCAASCloud(c *gc.C) {
 
 // SeedDatabase the database with a supplied controller config, and dummy
 // cloud and dummy credentials.
-func SeedDatabase(c *gc.C, runner database.TxnRunner, controllerConfig controller.Config) {
+func SeedDatabase(c *gc.C, runner coredatabase.TxnRunner, controllerConfig controller.Config) {
```
Comment (tlm) [inline juju/testing/apiserver.go:569]: Fix this import up.

File: juju/testing/apiserver.go [hunk 8]
```diff
@@ -565,19 +566,19 @@ func (s *ApiServerSuite) SeedCAASCloud(c *gc.C) {
...
 	err := controllerconfigbootstrap.InsertInitialControllerConfig(controllerConfig)(context.Background(), runner)
 	c.Assert(err, jc.ErrorIsNil)
 
 	SeedCloudCredentials(c, runner)
 }
 
-func SeedCloudCredentials(c *gc.C, runner database.TxnRunner) {
-	err := cloudbootstrap.InsertInitialControllerCloud(DefaultCloud)(context.Background(), runner)
+func SeedCloudCredentials(c *gc.C, runner coredatabase.TxnRunner) {
```
Comment (tlm) [inline juju/testing/apiserver.go:576]: Fix this import up.

File: juju/testing/apiserver.go [hunk 9]
```diff
@@ -623,7 +624,7 @@ func (s stubDBGetter) GetWatchableDB(namespace string) (changestream.WatchableDB
 }
 
 type stubWatchableDB struct {
-	database.TxnRunner
+	coredatabase.TxnRunner
```
Comment (tlm) [inline juju/testing/apiserver.go:627]: Fix this import up.

File: agent/agentbootstrap/bootstrap.go [hunk 10]
```diff
@@ -201,22 +204,49 @@ func (b *AgentBootstrap) Initialize(ctx stdcontext.Context) (_ *state.Controller
...
+	}
+
 	// Add the controller model cloud and credential to the database.
 	cloudCred, cloudCredTag, err := b.getCloudCredential()
 	if err != nil {
 		return nil, errors.Annotate(err, "getting cloud credentials from args")
 	}
-	ctrlCloud := stateParams.ControllerCloud
-	ctrlCloud.IsControllerCloud = true
+	if cloudCredTag.Id() != "" {
```
Comment (wallyworld) [inline agent/agentbootstrap/bootstrap.go:217]: There should always be a credential for the controller cloud - this should never be empty.

File: agent/agentbootstrap/bootstrap.go [hunk 11]
```diff
@@ -201,22 +204,49 @@ func (b *AgentBootstrap) Initialize(ctx stdcontext.Context) (_ *state.Controller
...
+		Type:        controllerModelType,
+	}
+	dqliteBootstrapOps = append(
+		dqliteBootstrapOps,
+		modelbootstrap.CreateModel(controllerUUID, controllerModelArgs),
+	)
 
 	if err := b.bootstrapDqlite(
-		ctx,
+		stdcontext.TODO(),
```
Comment (wallyworld) [inline agent/agentbootstrap/bootstrap.go:245]: Why replace the passed in context with a TODO?

File: domain/cloud/state/state.go [hunk 12]
```diff
@@ -294,11 +294,11 @@ ON CONFLICT(region_uuid, key) DO UPDATE
 
 func loadClouds(ctx context.Context, tx *sql.Tx, name string) ([]cloud.Cloud, error) {
 	// First load the basic cloud info and auth types.
+	// TODO: needs controller cloud
```
Comment (wallyworld) [inline domain/cloud/state/state.go:297]: If "is_controller_cloud" is not set, bootstrap on k8s will break. 
Did we really want this change in this PR? We can't leave stuff broken - we should do the fix in a separate PR where all the relevant work can be completed.

File: domain/credential/types.go [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+	Cloud string
+
+	// Owner is the owner of the credential.
+	Owner string
+
+	// Name is the name of the credential.
+	Name string
+}
+
+func IdFromTag(tag names.CloudCredentialTag) Id {
```
Comment (wallyworld) [inline domain/credential/types.go:25]: Comment

File: domain/model/bootstrap/bootstrap.go [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+	"context"
+	"database/sql"
+	"fmt"
+
+	"github.com/juju/juju/core/database"
+	"github.com/juju/juju/domain/model"
+	"github.com/juju/juju/domain/model/state"
+)
+
+// Create is responsible for making a new model with all of it's associated
```
Comment (wallyworld) [inline domain/model/bootstrap/bootstrap.go:16]: CreateModel

s/it's/its

File: domain/model/bootstrap/bootstrap.go [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+// metadata during the bootstrap process.
+func CreateModel(
+	uuid model.UUID,
+	args model.ModelCreationArgs,
+) func(context.Context, database.TxnRunner) error {
+	return func(ctx context.Context, db database.TxnRunner) error {
+		if err := args.Validate(); err != nil {
+			return fmt.Errorf("bootstrap model creation args: %w", err)
```
Comment (wallyworld) [inline domain/model/bootstrap/bootstrap.go:24]: This should be "controller model creation args:"

File: domain/model/service/service.go [hunk 16]
```diff
@@ -8,17 +8,18 @@ import (
...
 	"github.com/juju/juju/domain/credential"
 	"github.com/juju/juju/domain/model"
-	modelerrors "github.com/juju/juju/domain/model/errors"
 )
 
 // State is the model state required by this service.
 type State interface {
-	// SetCloudCredential sets the cloud credential for the given mode.
-	SetCloudCredential(context.Context, model.UUID, credential.ID) error
+	// Create creates a new model with all of it's associated metadata.
```
Comment (wallyworld) [inline domain/model/service/service.go:18]: s/it's/its

File: domain/model/service/service.go [hunk 17]
```diff
@@ -34,24 +35,46 @@ func NewService(st State) *Service {
...
+// - errors.ModelNotFound: When the model does not exist.
+// - errors.NotFound: When the cloud or credential cannot be found.
+// - errors.NotValid: When the cloud credential is not of the same cloud as the
+// model.
+func (s *Service) UpdateCloudCredential(
+	ctx context.Context,
+	uuid model.UUID,
+	id credential.Id,
+) error {
+	if id.IsZero() {
```
Comment (wallyworld) [inline domain/model/service/service.go:75]: This should be Validate()

File: domain/model/state/state.go [hunk 18]
```diff
@@ -32,20 +32,252 @@ func NewState(
...
+	})
+}
+
+// Create is responsible for creating a new model from start to finish. It will
+// register the model existence and associate all of the model metadata.
+// If a model already exists with the same name and owner then an error
+// satisfying modelerrors.AlreadyExists will be returned.
+// If the model type is not found then an error satisfying errors.NotSupported
+// will be returned.
+func Create(
```
Comment (wallyworld) [inline domain/model/state/state.go:62]: Why are these exported?

**Review by wallyworld — COMMENTED** [review]
Some initial comments, review still in progress.

File: domain/model/service/service.go (hunk 17 above)
Comment (tlm) [reply domain/model/service/service.go:75]: Yes, yes it should.

File: domain/model/bootstrap/bootstrap.go (hunk 15 above)
Comment (tlm) [reply domain/model/bootstrap/bootstrap.go:24]: Why? There is nothing controller specific here. If it's related to the controller then caller can annotate that further.

File: domain/cloud/state/state.go (hunk 12 above)
Comment (tlm) [reply domain/cloud/state/state.go:297]: I'll work on this. But we can ask this question we just need to wire it up a little better.

File: agent/agentbootstrap/bootstrap.go (hunk 11 above)
Comment (tlm) [reply agent/agentbootstrap/bootstrap.go:245]: was a bad rebase I just missed it in the dirty conflicts.

File: agent/agentbootstrap/bootstrap.go (hunk 10 above)
Comment (tlm) [reply agent/agentbootstrap/bootstrap.go:217]: This isn't the conversation we had in London. I can fix up.

File: internal/database/txn/errors_test.go [hunk 19]
```diff
@@ -4,9 +4,9 @@
 package txn_test
 
 import (
+	"github.com/juju/errors"
 	"github.com/juju/testing"
 	"github.com/mattn/go-sqlite3"
-	"github.com/pkg/errors"
```
Comment (SimonRichardson) [inline internal/database/txn/errors_test.go:9]: Good spot!

**Review by SimonRichardson — COMMENTED** [review]
The first Q&A looks good.

File: domain/model/bootstrap/bootstrap.go (hunk 15 above)
Comment (wallyworld) [reply domain/model/bootstrap/bootstrap.go:31]: The model created at bootstrap time IS the "controller" model - literally. That's its name. It's not the "bootstrap" model. "bootstrap" is the operation we are performing, and the end result is that the model called "controller" is created.

File: domain/modelmanager/state/state.go [hunk 20]
```diff
@@ -46,9 +48,12 @@ func (s *State) Create(ctx context.Context, uuid model.UUID) error {
 func Create(ctx context.Context, uuid model.UUID, tx *sql.Tx) error {
 	stmt := "INSERT INTO model_list (uuid) VALUES (?);"
 	result, err := tx.ExecContext(ctx, stmt, uuid)
-	if err != nil {
+	if jujudb.IsErrConstraintPrimaryKey(err) {
+		return fmt.Errorf("model for uuid %q %w", uuid, modelerrors.AlreadyExists)
```
Comment (manadart) [inline domain/modelmanager/state/state.go:52]: We haven't discussed domain-sourced errors yet, but I'm thinking they make sense.

File: domain/cloud/state/state.go (hunk 12 above)
Comment (manadart) [reply domain/cloud/state/state.go:297]: This is the right move though. The relational algebra has no way to constrain having is_controller_cloud=true for a cloud that is not used by the controller.

So always answering that question by SQL is correct.

File: domain/credential/state/types.go [hunk 21]
```diff
@@ -49,6 +49,17 @@ type Credential struct {
 	InvalidReason string `db:"invalid_reason"`
 }
 
+type CredentialAttribute struct {
```
Comment (manadart) [inline domain/credential/state/types.go:52]: Comment for the type.

Comment (manadart) [reply domain/credential/state/types.go:52]: Do these ever need exporting? They shouldn't really escape the state sub-packages.

File: domain/credential/types.go [hunk 22]
```diff
@@ -0,0 +1,77 @@
...
+
+import (
+	"fmt"
+
+	"github.com/juju/errors"
+	"github.com/juju/names/v4"
+)
+
+// Id represents the id of a cloud credential.
+type Id struct {
```
Comment (manadart) [inline domain/credential/types.go:14]: Isn't `ID` the idiom in Go?

File: domain/model/state/state.go [hunk 23]
```diff
@@ -32,20 +32,220 @@ func NewState(
...
+		return nil
+	})
+}
+
+// UpdateCloudCredential is responsible for updating the cloud credential in use
+// by model. If the cloud credential is not found an error that satisfies
+// errors.NotFound is returned.
+// If the credential being updated to is not of the same cloud that is currently
+// set for the model then an error that satisfies errors.NotValid is returned.
+func (s *State) UpdateCloudCredential(
```
Comment (manadart) [inline domain/model/state/state.go:223]: How about just `UpdateCredential`?

File: domain/model/state/state.go [hunk 24]
```diff
@@ -32,20 +32,220 @@ func NewState(
...
+			errors.NotFound, input.CloudRegion, input.Cloud,
+		)
+	} else if err != nil {
+		return fmt.Errorf(
+			"getting cloud region uuid for %q and cloud %q: %w",
+			input.CloudRegion, input.Cloud, err,
+		)
+	}
+
+	modelTypeStmt := "SELECT id from model_type WHERE type = ?"
```
Comment (manadart) [inline domain/model/state/state.go:142]: This is pretty easy to do in the single statement. Then you don't have to pull anything off the wire and into a result set.

Have a look at `ExtendLease` in the lease state.

File: domain/modelmanager/bootstrap/bootstrap.go [hunk 25]
```diff
@@ -6,18 +6,29 @@ package bootstrap
...
+// RegisterModel is responsible for registering the existence of a model in Juju.
+// The act of registering a model does not mean the model exists in Juju terms
+// for users to deploy recourse's to, but registers the fact that a model uuid
```
Comment (manadart) [inline domain/modelmanager/bootstrap/bootstrap.go:20]: "resources"?

**Review by manadart — COMMENTED** [review]
This is looking good.

File: domain/model/state/state.go (hunk 24 above)
Comment (tlm) [reply domain/model/state/state.go:142]: Thanks, I had it this way and perfectly happy to do it this way. The reason I opted over to this path was so that we can produce a better error message. But in hind sigh the validation of the creation args will catch it.

File: domain/model/bootstrap/bootstrap.go (hunk 15 above)
Comment (tlm) [reply domain/model/bootstrap/bootstrap.go:31]: Have removed the work "bootstrap"

File: domain/modelmanager/state/state.go (hunk 20 above)
Comment (tlm) [reply domain/modelmanager/state/state.go:52]: I wouldn't say this work is the perfect case of it yet. I am struggling with the balance.

But what I am trying to achieve in creating a model is there can be a lot of touch points for NotFound. The generic error is far too big of a funnel for what wasn't found. Trying it to narrow down what wasn't found for the consumer.

File: domain/schema/controller.go [hunk 26]
```diff
@@ -318,21 +317,47 @@ CREATE TABLE model_list (
...
-        REFERENCES model_list(uuid),
-    CONSTRAINT fk_model_metadata_cloud
-        FOREIGN KEY (cloud_uuid)
-        REFERENCES cloud(uuid),
-    CONSTRAINT fk_model_metadata_cloud_credential
-        FOREIGN KEY (cloud_credential_uuid)
-        REFERENCES cloud_credential(uuid)
-);`)
+CREATE UNIQUE INDEX idx_model_metadata_name_owner
+ON model_metadata (name, owner_uuid);
```
[the remark this answers is not in the document]
Comment (tlm) [reply domain/schema/controller.go:359]: We do allow this and we don't handle it for the Kubernetes case today. This is an ongoing bug that we need to deal with but we aren't introducing a new bug as part of this work we just aren't fixing it yet.

Ideally we can have conversation and design on how to handle this in the Kubernetes space.

File: domain/model/state/state.go [hunk 27]
```diff
@@ -32,11 +32,252 @@ func NewState(
...
+	}
+
+	if input.CloudRegion != "" {
+		err := setCloudRegion(ctx, uuid, input.CloudRegion, tx)
+		if err != nil {
+			return err
+		}
+	}
+
+	if !input.Credential.IsZero() {
```
Comment (wallyworld) [inline domain/model/state/state.go:167]: We should check for validity here, the reason being that we should error if a partially filled in credential is passed in. If it is partially filled in, we'll never attempt to update but will also never surface the required error.

File: domain/model/state/state.go [hunk 28]
```diff
@@ -32,11 +32,252 @@ func NewState(
...
+			if err := rows.Scan(&t); err != nil {
+				return fmt.Errorf("building model type: %w", err)
+			}
+			rval = append(rval, t)
+		}
+		return nil
+	})
+}
+
+// setCloudRegion is responsible for setting a models cloud region. This
```
Comment (wallyworld) [inline domain/model/state/state.go:206]: model's

File: domain/model/state/state.go [hunk 29]
```diff
@@ -58,39 +315,42 @@ AND cloud_credential.name = ?
...
-ON CONFLICT(model_uuid) DO UPDATE
-SET cloud_uuid = excluded.cloud_uuid,
-    cloud_credential_uuid = excluded.cloud_credential_uuid
-WHERE model_uuid = excluded.model_uuid
+UPDATE model_metadata
+SET cloud_credential_uuid = ?
+WHERE model_uuid = ?
+AND cloud_region_uuid IN (SELECT uuid
+                          FROM cloud_region
+						  WHERE cloud_uuid = ?)
```
Comment (wallyworld) [inline domain/model/state/state.go:323]: format or github glitch?

File: domain/schema/controller.go [hunk 30]
```diff
@@ -318,21 +317,51 @@ CREATE TABLE model_list (
...
+CREATE TABLE model_type (
+    id INT PRIMARY KEY,
+    type TEXT NOT NULL
+);
+
+CREATE UNIQUE INDEX idx_model_type_type
+ON model_type (type);
+
+INSERT INTO model_type VALUES
+    (0, 'none'),
```
Comment (wallyworld) [inline domain/schema/controller.go:329]: none is never a valid model type - it must always be set to iaas or caas (or whatever else)
so i don't think we want the 0 value here

**Review by wallyworld — APPROVED** [review]
Great progress. Just a few small things before landing.

File: domain/model/state/state.go (hunk 29 above)
Comment (tlm) [reply domain/model/state/state.go:323]: format.

File: domain/model/state/state.go (hunk 27 above)
Comment (wallyworld) [reply domain/model/state/state.go:167]: Checked in services layer, ignore me

File: domain/schema/controller.go (hunk 30 above)
Comment (tlm) [reply domain/schema/controller.go]: Done & Done

File: domain/model/state/state.go [hunk 31]
```diff
@@ -32,11 +32,285 @@ func NewState(
...
+		return fmt.Errorf(
+			"model %q already has a cloud region set%w",
+			uuid,
+			errors.Hide(errors.AlreadyExists),
+		)
+	}
+	return nil
+}
+
+// updateCredentialUpdateCredential is responsible for updating the cloud credential in use
```
Comment (wallyworld) [inline domain/model/state/state.go:308]: FIx me
